rust/rope/src/multiset.rs

```diff
@@ -0,0 +1,672 @@
+
+    #[test]
+    fn test_mapper() {
+        let substr = "469ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwz";
+        let s = find_deletions(substr, TEST_STR);
+        let mut m = s.mapper(CountMatcher::NonZero);
+        // subset is {0123 5 78 xy}
+        assert_eq!(0, m.doc_index_to_subset(0));
+        assert_eq!(2, m.doc_index_to_subset(2));
+        assert_eq!(2, m.doc_index_to_subset(2));
+        assert_eq!(3, m.doc_index_to_subset(3));
+        assert_eq!(4, m.doc_index_to_subset(4)); // not in subset
+        assert_eq!(4, m.doc_index_to_subset(5));
+        assert_eq!(5, m.doc_index_to_subset(7));
+        assert_eq!(6, m.doc_index_to_subset(8));
+        assert_eq!(6, m.doc_index_to_subset(8));
+        assert_eq!(8, m.doc_index_to_subset(60));
+        assert_eq!(9, m.doc_index_to_subset(61)); // not in subset
+        assert_eq!(9, m.doc_index_to_subset(62)); // not in subset
+    }
+
+    #[test]
+    #[should_panic(expected = "non-decreasing")]
+    fn test_mapper_requires_non_decreasing() {
+        let substr = "469ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvw";
+        let s = find_deletions(substr, TEST_STR);
+        let mut m = s.mapper(CountMatcher::NonZero);
+        m.doc_index_to_subset(0);
+        m.doc_index_to_subset(2);
+        m.doc_index_to_subset(1);
+    }
+
+    #[test]
+    fn union() {
+        let s1 = find_deletions("024AEGHJKNQTUWXYZabcfgikqrvy", TEST_STR);
+        let s2 = find_deletions("14589DEFGIKMOPQRUXZabcdefglnpsuxyz", TEST_STR);
+        assert_eq!("4EGKQUXZabcfgy", s1.union(&s2).delete_from_string(TEST_STR));
+    }
+
+    fn transform_case(str1: &str, str2: &str, result: &str) {
+        let s1 = find_deletions(str1, TEST_STR);
+        let s2 = find_deletions(str2, str1);
+        let s3 = s2.transform_expand(&s1);
+        let str3 = s3.delete_from_string(TEST_STR);
+        assert_eq!(result, str3);
+        assert_eq!(str2, s3.transform_shrink(&s1).delete_from_string(&str3));
+        assert_eq!(str2, s2.transform_union(&s1).delete_from_string(TEST_STR));
+    }
+
+    #[test]
+    fn transform() {
+        transform_case("02345678BCDFGHKLNOPQRTUVXZbcefghjlmnopqrstwx", "027CDGKLOTUbcegopqrw",
+            "01279ACDEGIJKLMOSTUWYabcdegikopqruvwyz");
+        transform_case("01234678DHIKLMNOPQRUWZbcdhjostvy", "136KLPQZvy",
+            "13569ABCEFGJKLPQSTVXYZaefgiklmnpqruvwxyz");
+        transform_case("0125789BDEFIJKLMNPVXabdjmrstuwy", "12BIJVXjmrstu",
+            "12346ABCGHIJOQRSTUVWXYZcefghijklmnopqrstuvxz");
+        transform_case("12456789ABCEFGJKLMNPQRSTUVXYadefghkrtwxz", "15ACEFGKLPRUVYdhrtx",
+            "0135ACDEFGHIKLOPRUVWYZbcdhijlmnopqrstuvxy");
+        transform_case("0128ABCDEFGIJMNOPQXYZabcfgijkloqruvy", "2CEFGMZabijloruvy",
+            "2345679CEFGHKLMRSTUVWZabdehijlmnoprstuvwxyz");
+        transform_case("01245689ABCDGJKLMPQSTWXYbcdfgjlmnosvy", "01245ABCDJLQSWXYgsv",
+            "0123457ABCDEFHIJLNOQRSUVWXYZaeghikpqrstuvwxz");
+    }
+
+    use subset;
```

Comment:
New test code starts here, all previous tests are unchanged except for updating based on the API changes.